### PR TITLE
[SPARK-16350][SQL] Fix support for incremental planning in wirteStream.foreach()

### PR DESCRIPTION
**Pull request by lw-lin**
## What changes were proposed in this pull request?

There are cases where `complete` output mode does not output updated aggregated value; for details please refer to [SPARK-16350](https://issues.apache.org/jira/browse/SPARK-16350).

The cause is that, as we do `data.as[T].foreachPartition { iter => ... }` in `ForeachSink.addBatch()`, `foreachPartition()` does not support incremental planning for now.

This patches makes `foreachPartition()` support incremental planning in `ForeachSink`, by making a special version of `Dataset` with its `rdd()` method supporting incremental planning.
## How was this patch tested?

Added a unit test which failed before the change

### CODE REVIEW

**Comment by lw-lin** [comment]
Jenkins retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala [hunk 1]
```diff
@@ -35,35 +35,109 @@ class ForeachSinkSuite extends StreamTest with SharedSQLContext with BeforeAndAf
...
         ForeachSinkSuite.Process(value = 4),
         ForeachSinkSuite.Close(None)
       )
 
-      val allEvents = ForeachSinkSuite.allEvents()
+      var allEvents = ForeachSinkSuite.allEvents()
+      assert(allEvents.size === 2)
+      assert {
+        allEvents === Seq(expectedEventsForPartition0, expectedEventsForPartition1) ||
+          allEvents === Seq(expectedEventsForPartition1, expectedEventsForPartition0)
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: `should contain theSameElementsAs`? See http://www.scalatest.org/user_guide/using_matchers#workingWithAggregations

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala [hunk 2]
```diff
@@ -35,35 +35,109 @@ class ForeachSinkSuite extends StreamTest with SharedSQLContext with BeforeAndAf
...
+        ForeachSinkSuite.Process(value = 6),
+        ForeachSinkSuite.Process(value = 8),
+        ForeachSinkSuite.Close(None)
+      )
+
+      allEvents = ForeachSinkSuite.allEvents()
       assert(allEvents.size === 2)
       assert {
         allEvents === Seq(expectedEventsForPartition0, expectedEventsForPartition1) ||
           allEvents === Seq(expectedEventsForPartition1, expectedEventsForPartition0)
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: Same as above

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala [hunk 3]
```diff
@@ -35,35 +35,109 @@ class ForeachSinkSuite extends StreamTest with SharedSQLContext with BeforeAndAf
...
+
+  test("foreach() with `complete` output mode") {
+    withTempDir { checkpointDir =>
+      val input = MemoryStream[Int]
+
+      val query = input.toDS()
+        .groupBy().count().as[Long].map(_.toInt)
+        .writeStream
+        .option("checkpointLocation", checkpointDir.getCanonicalPath)
+        .outputMode("complete")
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: Are really output modes strings? No enums or similar more type-safe values?

**Comment by lw-lin** [comment]
Jenkins retest this please

**Comment by lw-lin** [comment]
@zsxwing could you take a look at this? Thanks!

**Comment by lw-lin** [comment]
@zsxwing could you take a look at this when you have time? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 4]
```diff
@@ -155,7 +155,7 @@ private[sql] object Dataset {
 class Dataset[T] private[sql](
     @transient val sparkSession: SparkSession,
     @DeveloperApi @transient val queryExecution: QueryExecution,
-    encoder: Encoder[T])
+    val encoder: Encoder[T])
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: I'm not sure we want to make this publicly available.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala [hunk 5]
```diff
@@ -30,7 +32,42 @@ import org.apache.spark.sql.{DataFrame, Encoder, ForeachWriter}
 class ForeachSink[T : Encoder](writer: ForeachWriter[T]) extends Sink with Serializable {
 
   override def addBatch(batchId: Long, data: DataFrame): Unit = {
-    data.as[T].foreachPartition { iter =>
+    // TODO: Refine this method when SPARK-16264 is resolved; see comments below.
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: Could we just use the QueryExecution of the dataframe that is passed in, instead of creating a new one?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala [hunk 6]
```diff
@@ -30,7 +32,42 @@ import org.apache.spark.sql.{DataFrame, Encoder, ForeachWriter}
...
+    // does not support `IncrementalExecution`.
+    //
+    // So as a provisional fix, below we've made a special version of `Dataset` with its `rdd()`
+    // method supporting incremental planning. But in the long run, we should generally make newly
+    // created Datasets use `IncrementalExecution` where necessary (which is SPARK-16264 tries to
+    // resolve).
+
+    val dataAsT = data.as[T]
+    val datasetWithIncrementalExecution =
+      new Dataset(data.sparkSession, dataAsT.logicalPlan, dataAsT.encoder) {
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: `dataAsT` can be removed. You can use `implicitly[Encoder[T]]` to get the encoder. Please also revert the change to Dataset.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala (hunk 1 above)
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: yeah, or just `allEvents.toSet === Set(expectedEventsForPartition0, expectedEventsForPartition1)`

**Comment by zsxwing** [comment]
@lw-lin thanks for fixing it. Just some minor comments. Once you fix it and tests pass, I will merge it.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 4 above)
Comment (lw-lin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: this change had been reverted; thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala (hunk 6 above)
Comment (lw-lin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: fixed; thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala (hunk 1 above)
Comment (lw-lin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: fixed. @jaceklaskowski @zsxwing thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala (hunk 3 above)
Comment (lw-lin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: fixed; thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala (hunk 5 above)
Comment (lw-lin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: The logical plan of the created `IncrementalExecution` has to be `deserialized`[#L55](https://github.com/apache/spark/pull/14030/files#diff-98acda846a9dd63efc42e0957594e05dR55),
so we should not re-use the `QueryExecution` passed in?

**Comment by lw-lin** [comment]
Updated. @zsxwing could you take another look?

**Comment by zsxwing** [comment]
LGTM. Merging to master and 2.0.
